lib/linklist.h

```diff
@@ -43,6 +43,12 @@ struct list {
 	/* invariant: count is the number of listnodes in the list */
 	unsigned int count;
 
+	unsigned int flags;
```

Comment:
Short story - I need DLLs.
Long story - I need to link the same entry to multiple lists for efficient walks. At the same time I do NOT want to find/lookup and del i.e. as I already have the entry I just need to unlink (without find) from all of these lists. So the right construct for me is DLL (efficient walks/efficient dels with the least amount of memory per-entry). And DLLs are already used for this exact purpose in the EVPN code a lot i.e. my usage is consistent.

why did I make the change? I need to unlink from lists without lookup, which means (with the way the current lib is written) I need to store the dll_node pointer(s) in the entry - that is 8 useless bytes; and as i link each entry to multiple lists it is N*8 useless bytes. Instead if i had the dll_node memory in the entry itself I save [entries*lists*8 bytes]. This is a simple yet needed optimization.
A dll_node should be no more than 2*8bytes. Everything else is fat that can/must be trimmed.

PS: Off course there are cases where an entry can be linked to a list to temporarily i.e. you do not want to permanently burn dll_node memory per-entry. for such cases you continue to let the DLL lib manage the dll_node mem. That option will continue to stay.

I am planning on some more optimizations to the DLL lib. I will send that as a separate PR. It will take me less time to write a new DLL lib. But @donaldsharp  has rightly taken a stand against the proliferation of DLLs. Plus I want to use the same DLL lib as the existing EVPN code (consistency etc.).

Reply:
Thank you for your explanation.

Quick clarification questions:
1. Does DLL mean Doubly Linked List?
2. Are there situations where you don't use a list node?

Here are some information for question (2):
If you always have the entry "linked" to all lists, I'd suggest you to move to `SLISTS` or `TAILQ`. The linked list implementation using separated list nodes is a bit less efficient than using `SLISTS` or `TAILQ`, starting with the data structure:

```c
struct listnode {
        struct listnode *next;
        struct listnode *prev;

        /* private member, use getdata() to retrieve, do not access directly */
        void *data;
};
```

vs

```c
#define SLIST_ENTRY(type)                                                      \
        struct {                                                               \
                struct type *sle_next; /* next element */                      \
        }

/* or */

#define TAILQ_ENTRY(type)                                                      \
        struct {                                                               \
                struct type *tqe_next;  /* next element */                     \
                struct type **tqe_prev; /* address of previous next element */ \
        }
```

> In both cases you always save 1 pointer at least - the `data` pointer - and you get lists with quick removal (see comments below)

The second biggest difference is how those are allocated:

1. `lib/linkedlist.h` will use `(m|c)alloc` for each entry providing some data safety, but fragmenting memory and using more memory
2. `lib/openbsd-queue.h` will be allocated with your data structure saving a `(c|m)alloc` and using less memory

I don't mind your approach, but I think you could've avoided these library changes by just using a different list implementation.

I also think that the `SLIST`/`TAILQ` seem to be appropriated for your usage:

```c
/*
 * If you can afford only removing from head or when iterating, SLIST might work for you:
 */

#define SLIST_REMOVE_AFTER(elm, field)                                         \
        do {                                                                   \
                (elm)->field.sle_next = (elm)->field.sle_next->field.sle_next; \
        } while (0)

#define SLIST_REMOVE_HEAD(head, field)                                         \
        do {                                                                   \
                (head)->slh_first = (head)->slh_first->field.sle_next;         \
        } while (0)

/* or */

#define TAILQ_REMOVE(head, elm, field)                                         \
        do {                                                                   \
                if (((elm)->field.tqe_next) != NULL)                           \
                        (elm)->field.tqe_next->field.tqe_prev =                \
                                (elm)->field.tqe_prev;                         \
                else                                                           \
                        (head)->tqh_last = (elm)->field.tqe_prev;              \
                *(elm)->field.tqe_prev = (elm)->field.tqe_next;                \
                _Q_INVALIDATE((elm)->field.tqe_prev);                          \
                _Q_INVALIDATE((elm)->field.tqe_next);                          \
        } while (0)
```

I'm just writing this comment to give you options before you go optimizing your current approach.

Reply:
Sure, I hear you. SLISTs and TAILq are not efficient for deletion, so DLL (yes, those are doubly linked lists) are most suitable for the misc EVPN purposes.